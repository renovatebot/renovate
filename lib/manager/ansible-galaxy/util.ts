export const newBlockRegEx = /^\s*-\s*((\w+):\s*(.*))$/;
export const blockLineRegEx = /^\s*((\w+):\s*(.*))$/;
export const galaxyDepRegex = /[\w-]+\.[\w-]+/;
export const dependencyRegex = /^dependencies:/;
export const galaxyRegEx =
  /^\s+(?<lookupName>[\w.]+):\s*["'](?<version>.+)["']\s*/;
export const nameMatchRegex =
  /(?<source>((git\+)?(?:(git|ssh|https?):\/\/)?(.*@)?(?<hostname>[\w.-]+)(?:(:\d+)?\/|:))(?<depName>[\w./-]+)(?:\.git)?)(,(?<version>[\w.]*))?/;
