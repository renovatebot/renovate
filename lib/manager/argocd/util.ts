export const keyValueExtractionRegex =
  /^\s*(?<key>[^\s]+):\s+"?(?<value>[^"\s]+)"?\s*$/;
// looks for `apiVersion: argoproj.io/
export const fileTestRegex = /\s*apiVersion:\s*argoproj.io\/\s*/;
