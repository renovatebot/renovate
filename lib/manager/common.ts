import { ReleaseType } from 'semver';
import { RangeStrategy } from '../versioning';
import { ValidationMessage } from '../config/common';

export type Result<T> = T | Promise<T>;

export interface ManagerConfig {
  binarySource?: string;
  dockerUser?: string;
  localDir?: string;
  registryUrls?: string[];
}

export interface ManagerData<T> {
  managerData?: T;
}

export interface ExtractConfig extends ManagerConfig {
  endpoint?: string;
  global?: any;
  gradle?: { timeout?: number };
  aliases?: Record<string, string>;
  ignoreNpmrcFile?: boolean;

  skipInstalls?: boolean;
  versionScheme?: string;
}

export interface UpdateArtifactsConfig extends ManagerConfig {
  isLockFileMaintenance?: boolean;
  compatibility?: Record<string, string>;
  cacheDir?: string;
  postUpdateOptions?: string[];
  ignoreScripts?: boolean;
}

export interface PackageUpdateConfig {
  currentValue?: string;
  rangeStrategy?: RangeStrategy;
  supportPolicy?: string[];
}

export interface PackageUpdateResult {
  newValue: string[];
  newMajor: string;
  isRange: boolean;
  sourceUrl: string;
}

export interface RangeConfig<T = Record<string, any>> extends ManagerData<T> {
  composerJsonType?: 'composer-plugin' | 'library' | 'metapackage' | 'project';
  currentValue?: string;
  depName?: string;
  depType?: string;
  manager?: string;
  packageJsonType?: 'app' | 'library';
  rangeStrategy: RangeStrategy;
}

export interface NpmLockFiles {
  yarnIntegrity?: boolean;
  yarnLock?: string;
  packageLock?: string;
  shrinkwrapJson?: string;
  pnpmShrinkwrap?: string;
  npmLock?: string;
  lernaDir?: string;
}

export interface PackageFile<T = Record<string, any>>
  extends NpmLockFiles,
    ManagerData<T> {
  hasYarnWorkspaces?: boolean;
  internalPackages?: string[];
  compatibility?: Record<string, string>;
  datasource?: string;
  registryUrls?: string[];
  deps: PackageDependency[];
  ignoreNpmrcFile?: boolean;
  lernaClient?: string;
  lernaPackages?: string[];
  manager?: string;
  mavenProps?: Record<string, any>;
  npmrc?: string;
  packageFile?: string;
  packageJsonName?: string;
  packageJsonType?: 'app' | 'library';
  packageJsonVersion?: string;
  parent?: string;
  skipInstalls?: boolean;
  yarnrc?: string;
  yarnWorkspacesPackages?: string[] | string;
}

export interface Package<T> extends ManagerData<T> {
  currentValue?: string;
  currentDigest?: string;
  depName?: string;
  depNameShort?: string;
  depType?: string;
  fileReplacePosition?: number;
  groupName?: string;
  lineNumber?: number;
  lookupName?: string;
  repo?: string;
  target?: string;
  versionScheme?: string;

  // npm manager
  bumpVersion?: ReleaseType | string;
  npmPackageAlias?: boolean;
  packageJsonVersion?: string;
  gitRef?: boolean;
  sourceUrl?: string;
  githubRepo?: string;
  pinDigests?: boolean;
  currentRawValue?: string;
  major?: { enabled?: boolean };
  prettyDepType?: any;
}

export interface PackageDependency<T = Record<string, any>> extends Package<T> {
  warnings?: ValidationMessage[];
  commitMessageTopic?: string;
  currentDigestShort?: string;
  datasource?: string;
  deprecationMessage?: string;
  digestOneAndOnly?: boolean;
  fromVersion?: string;
  lockedVersion?: string;
  lookupType?: string;
  moduleName?: string;
  propSource?: string;
  registryUrls?: string[];
  rangeStrategy?: RangeStrategy;
  skipReason?: string;
  source?: string;
  sourceLine?: number;
  updates?: PackageUpdateResult[];
  versionLine?: number;
}

export interface Upgrade<T = Record<string, any>>
  extends Package<T>,
    NpmLockFiles {
  isLockfileUpdate?: boolean;
  currentRawValue?: any;
  checksumUrl?: string;
  currentVersion?: string;
  depGroup?: string;
  dockerRepository?: string;
  downloadUrl?: string;
  localDir?: string;
  name?: string;
  newDigest?: string;
  newFrom?: string;
  newMajor?: number;
  newValue?: string;
  newVersion?: string;
  packageFile?: string;
  rangeStrategy?: RangeStrategy;
  toVersion?: string;
  updateType?: string;
  version?: string;
}

export interface ArtifactError {
  lockFile?: string;
  stderr?: string;
}

export interface UpdateArtifactsResult {
  artifactError?: ArtifactError;
  file?: { name: string; contents: string };
}

export interface UpdateArtifact {
  packageFileName: string;
  updatedDeps: string[];
  newPackageFileContent: string;
  config: UpdateArtifactsConfig;
}
export interface ManagerApi {
  language?: string;
  supportsLockFileMaintenance?: boolean;

  extractAllPackageFiles?(
    config: ExtractConfig,
    files: string[]
  ): Result<PackageFile[] | null>;

  extractPackageFile?(
    content: string,
    packageFile?: string,
    config?: ExtractConfig
  ): Result<PackageFile | null>;

  getPackageUpdates?(
    config: PackageUpdateConfig
  ): Result<PackageUpdateResult[]>;

  getRangeStrategy(config: RangeConfig): RangeStrategy;

  updateArtifacts?(
    updateArtifact: UpdateArtifact
  ): Result<UpdateArtifactsResult[] | null>;

  updateDependency(
    fileContent: string,
    upgrade: Upgrade
  ): Result<string | null>;
}

// TODO: name and properties used by npm manager
export interface PostUpdateConfig extends ManagerConfig, Record<string, any> {
  cacheDir?: string;

  postUpdateOptions?: string[];
  skipInstalls?: boolean;

  platform?: string;
  upgrades?: Upgrade[];
  npmLock?: string;
  yarnLock?: string;
  branchName?: string;
  parentBranch?: string;
}
