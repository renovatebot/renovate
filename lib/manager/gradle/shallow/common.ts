export { MAVEN_REPO } from '../../../datasource/maven/common';

export const JCENTER_REPO = 'https://jcenter.bintray.com/';
export const GOOGLE_REPO = 'https://dl.google.com/android/maven2/';
export const GRADLE_PLUGIN_PORTAL_REPO = 'https://plugins.gradle.org/m2/';

// TODO: convert to types
// eslint-disable-next-line typescript-enum/no-enum
export enum TokenType {
  Space = 'space',
  LineComment = 'lineComment',
  MultiComment = 'multiComment',
  Newline = 'newline',

  Semicolon = 'semicolon',
  Colon = 'colon',
  Dot = 'dot',
  Comma = 'comma',
  Operator = 'operator',

  Assignment = 'assignment',

  Word = 'word',

  LeftParen = 'leftParen',
  RightParen = 'rightParen',

  LeftBracket = 'leftBracket',
  RightBracket = 'rightBracket',

  LeftBrace = 'leftBrace',
  RightBrace = 'rightBrace',

  SingleQuotedStart = 'singleQuotedStart',
  SingleQuotedFinish = 'singleQuotedFinish',

  DoubleQuotedStart = 'doubleQuotedStart',
  StringInterpolation = 'interpolation',
  IgnoredInterpolationStart = 'ignoredInterpolation',
  Variable = 'variable',
  DoubleQuotedFinish = 'doubleQuotedFinish',

  TripleSingleQuotedStart = 'tripleQuotedStart',
  TripleDoubleQuotedStart = 'tripleDoubleQuotedStart',
  TripleQuotedFinish = 'tripleQuotedFinish',

  Chars = 'chars',
  EscapedChar = 'escapedChar',
  String = 'string',

  UnknownFragment = 'unknownFragment',
}
