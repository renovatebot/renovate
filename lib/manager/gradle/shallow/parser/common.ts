import { GradleContext } from './types';

export const groupIdRegexPart =
  '(?<groupId>[a-zA-Z][-_a-zA-Z0-9]*(?:\\.[a-zA-Z0-9][-_a-zA-Z0-9]*?)*)';

export const artifactIdRegexPart = groupIdRegexPart.replace(
  '<groupId>',
  '<artifactId>'
);

export const versionRegexPart = '(?<version>[-.\\[\\](),a-zA-Z0-9+]+)';

export const dataTypeRegexPart =
  '(?:(?:@(?<dataType>[a-zA-Z][-_a-zA-Z0-9]*))?)';

export function cleanupContext(ctx: GradleContext): GradleContext {
  delete ctx.variableName;
  delete ctx.otherPackageFile;
  delete ctx.fileReplacePosition;
  delete ctx.groupId;
  delete ctx.artifactId;
  delete ctx.version;
  delete ctx.dataType;
  delete ctx.paramName;
  return ctx;
}
