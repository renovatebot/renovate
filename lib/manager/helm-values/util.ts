export type HelmDockerImageDependency = {
  registry?: string;
  repository: string;
  tag: string;
};

/**
 * This is a workaround helper to allow the usage of 'unknown' in
 * a type-guard function while checking that keys exist.
 *
 * @see https://github.com/microsoft/TypeScript/issues/21732
 * @see https://stackoverflow.com/a/58630274
 */
function hasKey<K extends string>(k: K, o: {}): o is { [_ in K]: {} } {
  return typeof o === 'object' && k in o;
}

/**
 * Type guard to determine whether a given partial Helm values.yaml object potentially
 * defines a Helm Docker dependency.
 *
 * There is no exact standard of how Docker dependencies are defined in Helm
 * values.yaml files (as of January 1st 2020), this function defines a
 * heuristic based on the most commonly used format in the stable Helm charts:
 *
 * image:
 *   repository: 'something'
 *   tag: v1.0.0
 */
export function matchesHelmValuesDockerHeuristic(
  parentKey: string,
  data: unknown
): data is HelmDockerImageDependency {
  return (
    parentKey === 'image' &&
    data &&
    typeof data === 'object' &&
    hasKey('repository', data) &&
    hasKey('tag', data)
  );
}
