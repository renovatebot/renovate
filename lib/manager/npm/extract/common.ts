import { PackageJson } from 'type-fest';

export interface NpmPackage extends PackageJson {
  workspaces?: any;
  renovate?: any;
  _from?: any;
  _args?: any;
  _id?: any;
}

export type LockFileEntry = Record<
  string,
  { version: string; integrity?: boolean }
>;
