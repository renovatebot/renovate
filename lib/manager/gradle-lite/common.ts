import { PackageDependency } from '../common';

export { MAVEN_REPO } from '../../datasource/maven/common';

export const JCENTER_REPO = 'https://jcenter.bintray.com/';
export const GOOGLE_REPO = 'https://dl.google.com/android/maven2/';

export interface ManagerData {
  fileReplacePosition: number;
  packageFile?: string;
}

export interface VariableData extends ManagerData {
  key: string;
  value: string;
}

export type PackageVariables = Record<string, VariableData>;
export type VariableRegistry = Record<string, PackageVariables>;

export enum TokenType {
  Space = 'space',
  LineComment = 'lineComment',
  MultiComment = 'multiComment',
  Newline = 'newline',

  Semicolon = 'semicolon',
  Colon = 'colon',
  Dot = 'dot',
  Comma = 'comma',
  Operator = 'operator',

  Assignment = 'assignment',

  Word = 'word',

  LeftParen = 'leftParen',
  RightParen = 'rightParen',

  LeftBracket = 'leftBracket',
  RightBracket = 'rightBracket',

  LeftBrace = 'leftBrace',
  RightBrace = 'rightBrace',

  SingleQuotedStart = 'singleQuotedStart',
  SingleQuotedFinish = 'singleQuotedFinish',

  DoubleQuotedStart = 'doubleQuotedStart',
  StringInterpolation = 'interpolation',
  IgnoredInterpolationStart = 'ignoredInterpolation',
  Variable = 'variable',
  DoubleQuotedFinish = 'doubleQuotedFinish',

  TripleSingleQuotedStart = 'tripleQuotedStart',
  TripleDoubleQuotedStart = 'tripleDoubleQuotedStart',
  TripleQuotedFinish = 'tripleQuotedFinish',

  Char = 'char',
  EscapedChar = 'escapedChar',
  String = 'string',

  UnknownLexeme = 'unknownChar',
  UnknownFragment = 'unknownFragment',
}

export interface Token {
  type: TokenType;
  value: string;
  offset: number;
}

export interface StringInterpolation extends Token {
  type: TokenType.StringInterpolation;
  children: Token[]; // Tokens inside double-quoted string that are subject of interpolation
  isComplete: boolean; // True if token has parsed completely
  isValid: boolean; // False if string contains something unprocessable
}

// Matcher on single token
export interface SyntaxMatcher {
  matchType: TokenType | TokenType[];
  matchValue?: string | string[];
  lookahead?: boolean;
  tokenMapKey?: string;
}

export type TokenMap = Record<string, Token>;

export interface SyntaxHandlerInput {
  packageFile: string;
  variables: PackageVariables;
  tokenMap: TokenMap;
}

export type SyntaxHandlerOutput = {
  deps?: PackageDependency<ManagerData>[];
  vars?: PackageVariables;
  urls?: string[];
} | null;

export interface SyntaxMatchConfig {
  matchers: SyntaxMatcher[];
  handler: (MatcherHandlerInput) => SyntaxHandlerOutput;
}
