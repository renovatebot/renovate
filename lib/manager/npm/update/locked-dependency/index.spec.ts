import { readFileSync } from 'fs';
import { resolve } from 'upath';
import * as httpMock from '../../../../../test/http-mock';
import { getName } from '../../../../../test/util';
import { clone } from '../../../../util/clone';
import { UpdateLockedConfig } from '../../../common';
import { updateLockedDependency } from '.';

const packageFileContent = readFileSync(
  resolve(__dirname, './__fixtures__/package.json'),
  'utf8'
);
const lockFileContent = readFileSync(
  resolve(__dirname, './__fixtures__/package-lock.json'),
  'utf8'
);

const acceptsJson = JSON.parse(
  readFileSync(resolve(__dirname, './__fixtures__/accepts.json'), 'utf8')
);

const expressJson = JSON.parse(
  readFileSync(resolve(__dirname, './__fixtures__/express.json'), 'utf8')
);

const mimeJson = JSON.parse(
  readFileSync(resolve(__dirname, './__fixtures__/mime.json'), 'utf8')
);

const serveStaticJson = JSON.parse(
  readFileSync(resolve(__dirname, './__fixtures__/serve-static.json'), 'utf8')
);

const sendJson = JSON.parse(
  readFileSync(resolve(__dirname, './__fixtures__/send.json'), 'utf8')
);

const typeIsJson = JSON.parse(
  readFileSync(resolve(__dirname, './__fixtures__/type-is.json'), 'utf8')
);

describe(getName(__filename), () => {
  describe('updateLockedDependency()', () => {
    let config: UpdateLockedConfig;
    beforeEach(() => {
      httpMock.setup();
      config = {
        packageFile: 'package.json',
        packageFileContent,
        lockFile: 'package-lock.json',
        lockFileContent,
        depName: 'some-dep',
        currentVersion: '1.0.0',
        newVersion: '1.0.1',
      };
    });
    afterEach(() => {
      httpMock.reset();
    });
    it('validates filename', async () => {
      expect(
        await updateLockedDependency({ ...config, lockFile: 'yarn.lock' })
      ).toBeNull();
    });
    it('validates versions', async () => {
      expect(
        await updateLockedDependency({
          ...config,
          newVersion: '^2.0.0',
        })
      ).toBeNull();
    });
    it('returns null for unparseable files', async () => {
      expect(
        await updateLockedDependency({ ...config, lockFileContent: 'not json' })
      ).toBeNull();
    });
    it('rejects lockFileVersion 2', async () => {
      expect(
        await updateLockedDependency({
          ...config,
          lockFileContent: lockFileContent.replace(': 1,', ': 2,'),
        })
      ).toBeNull();
    });
    it('returns null if no locked deps', async () => {
      expect(await updateLockedDependency(config)).toBeNull();
    });
    it('rejects null if no constraint found', async () => {
      expect(
        await updateLockedDependency({
          ...config,
          lockFileContent: lockFileContent.replace('1.0.0', '10.0.0'),
          depName: 'accepts',
          currentVersion: '10.0.0',
          newVersion: '11.0.0',
        })
      ).toBeNull();
    });
    it('remediates in-range', async () => {
      const res = await updateLockedDependency({
        ...config,
        depName: 'mime',
        currentVersion: '1.2.11',
        newVersion: '1.2.12',
      });
      expect(res['package-lock.json']).toContain('"mime":{"version":"1.2.12"');
    });
    it('fails to remediate if parent dep cannot support', async () => {
      const acceptsModified = clone(acceptsJson);
      acceptsModified.versions['2.0.0'] = {};
      httpMock
        .scope('https://registry.npmjs.org')
        .get('/accepts')
        .reply(200, acceptsModified);
      httpMock
        .scope('https://registry.npmjs.org')
        .get('/express')
        .reply(200, expressJson);
      const res = await updateLockedDependency({
        ...config,
        depName: 'accepts',
        currentVersion: '1.0.0',
        newVersion: '2.0.0',
      });
      expect(res).toBeNull();
    });
    it('remediates express', async () => {
      config.depName = 'express';
      config.currentVersion = '4.0.0';
      config.newVersion = '4.1.0';
      const res = await updateLockedDependency(config);
      expect(res['package.json']).toContain('"express": "4.1.0"');
      expect(res['package-lock.json']).toContain(
        '"express":{"version":"4.1.0"'
      );
    });
    it('remediates mime', async () => {
      config.depName = 'mime';
      config.currentVersion = '1.2.11';
      config.newVersion = '1.4.1';
      httpMock
        .scope('https://registry.npmjs.org')
        .get('/accepts')
        .reply(200, acceptsJson);
      httpMock
        .scope('https://registry.npmjs.org')
        .get('/express')
        .reply(200, expressJson);
      httpMock
        .scope('https://registry.npmjs.org')
        .get('/mime')
        .reply(200, mimeJson);
      httpMock
        .scope('https://registry.npmjs.org')
        .get('/send')
        .reply(200, sendJson);
      httpMock
        .scope('https://registry.npmjs.org')
        .get('/serve-static')
        .reply(200, serveStaticJson);
      httpMock
        .scope('https://registry.npmjs.org')
        .get('/type-is')
        .reply(200, typeIsJson);
      const res = await updateLockedDependency(config);
      expect(res['package-lock.json']).toContain('"mime":{"version":"1.4.1"');
      expect(res['package-lock.json']).toContain(
        '"express":{"version":"4.16.0"'
      );
      expect(res['package.json']).toContain('"express": "4.16.0"');
    });
  });
});
