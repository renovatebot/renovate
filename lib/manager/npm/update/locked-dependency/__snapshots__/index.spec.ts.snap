// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`manager/npm/update/locked-dependency/index updateLockedDependency() remediates mime 1`] = `
Array [
  Object {
    "headers": Object {
      "accept": "application/json",
      "accept-encoding": "gzip, deflate, br",
      "host": "registry.npmjs.org",
      "user-agent": "https://github.com/renovatebot/renovate",
    },
    "method": "GET",
    "url": "https://registry.npmjs.org/mime",
  },
  Object {
    "headers": Object {
      "accept": "application/json",
      "accept-encoding": "gzip, deflate, br",
      "host": "registry.npmjs.org",
      "user-agent": "https://github.com/renovatebot/renovate",
    },
    "method": "GET",
    "url": "https://registry.npmjs.org/send",
  },
  Object {
    "headers": Object {
      "accept": "application/json",
      "accept-encoding": "gzip, deflate, br",
      "host": "registry.npmjs.org",
      "user-agent": "https://github.com/renovatebot/renovate",
    },
    "method": "GET",
    "url": "https://registry.npmjs.org/type-is",
  },
  Object {
    "headers": Object {
      "accept": "application/json",
      "accept-encoding": "gzip, deflate, br",
      "host": "registry.npmjs.org",
      "user-agent": "https://github.com/renovatebot/renovate",
    },
    "method": "GET",
    "url": "https://registry.npmjs.org/serve-static",
  },
]
`;
